updater/updater.go

```diff
@@ -200,6 +251,144 @@ func upgradeIfRequired(checkURL string) (upgraded bool, err error) {
+	profileFile := filepath.Join(filepath.Dir(updaterBin), profileFileName)
+	return profileFile, nil
+}
+
+func validateProfile(prf *Profile) error {
+	isAlphaNumeric := regexp.MustCompile(`^[A-Za-z0-9]+$`).MatchString
+	// Should validate the id string and channel.
+	// For now lets assume id string is less than 256
+	// characters and alphanumeric.
+	// Channel to be alphanemeric and less than 10 characters.
+	if !isAlphaNumeric(prf.Id) ||
+		!isAlphaNumeric(prf.Channel) ||
+		len(prf.Id) > 256 ||
+		len(prf.Channel) > 10 {
+		return errors.New("Profile Id or Channel format is invalid.")
+	}
+	return nil
+}
+
+func generateRandomIdString() (string, error) {
+	nRand, err := rand.Int(rand.Reader, big.NewInt(math.MaxInt64))
+	if err != nil {
+		return "", err
+	}
+	return fmt.Sprintf("%s", nRand), nil
+}
+
+func loadProfile(prf *Profile) (err error) {
+	var fn string
+	var data []byte
+	if fn, err = getProfileFileName(); err != nil {
+		return err
+	}
+	if data, err = ioutil.ReadFile(fn); err != nil {
+		return err
+	}
+	if err = json.Unmarshal(data, prf); err != nil {
+		return err
+	}
+	return validateProfile(prf)
+}
+
+func saveProfile(prf *Profile) (err error) {
+	var fn string
+	var data []byte
+	if fn, err = getProfileFileName(); err != nil {
+		return err
+	}
+	if data, err = json.Marshal(prf); err != nil {
+		return err
+	}
+	return ioutil.WriteFile(fn, data, 0600)
+}
+
+func setRandomProfileId() int {
+	strRand, err := generateRandomIdString()
+	if err != nil {
+		fmt.Errorf("Error: %v\n", err)
+		return 1
+	}
+	prf := Profile{
+		Id:      strRand,
+		Channel: valChannelStable,
+	}
+	if err = saveProfile(&prf); err != nil {
+		fmt.Errorf("Error: %v\n", err)
+		return 1
+	}
+	return 0
+}
+
+func setProfileId(id string) int {
+	prf := Profile{}
+	err := loadProfile(&prf)
+	prf.Id = id
+	if err != nil {
+		// Profile load failed. Doesn't exist or corrupted.
+		// Set channel as well.
+		prf.Channel = valChannelStable
+	}
+	if err = saveProfile(&prf); err != nil {
+		fmt.Errorf("Error: %v.\n", err)
+		return 1
+	}
+	return 0
+}
+
+func setProfileChannel(channel string) int {
+	prf := Profile{}
+	err := loadProfile(&prf)
+	prf.Channel = channel
+	if err != nil {
+		// Profile load failed. Doesn't exist or corrupted.
+		// Set id as well.
+		strRand, errRand := generateRandomIdString()
+		if errRand != nil {
+			fmt.Errorf("Error: %v\n", errRand)
+			return 1
+		}
+		prf.Id = strRand
+	}
+	if err = saveProfile(&prf); err != nil {
+		fmt.Errorf("Error: %v.\n", err)
+		return 1
+	}
+	return 0
+}
+
+func addIdProfileToRequestHeader(req *http.Request) {
+	// Best effort to add identity profile to header.
+	// Errors are logged allowing normal operation.
+
+	// Add timezone. Gives a broad geo location.
+	t := time.Now()
+	zone, _ := t.Zone()
+	req.Header.Set(timezoneHeaderStr, zone)
+
+	// Add profile.
+	prf := Profile{}
+	if err := loadProfile(&prf); err != nil {
+		fmt.Printf("Couldn't load profile: %v.\n", err)
+		return
+	}
+	if len(prf.Id) > 0 {
```

Comment:
`prf.Id != ""` bit nicer to read?

Reply:
Debatable, but if that's the go way of doing things I'll change.